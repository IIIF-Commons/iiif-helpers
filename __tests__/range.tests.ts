import tableOfContentManifests from '../fixtures/cookbook/toc.json';
import {
  Vault,
  getValue,
  rangeToTableOfContentsTree,
  rangesToTableOfContentsTree,
  RangeTableOfContentsNode,
} from '../src';
import { ManifestNormalized } from '@iiif/presentation-3-normalized';
import tableOfContentsAvManifest from '../fixtures/cookbook/toc-av.json';
import wellcomeRange from '../fixtures/presentation-2/wellcome-range.json';
import blManifest from '../fixtures/presentation-3/bl-av-manifest.json';
import invariant from 'tiny-invariant';

// Test utility.
// Render range in ascii with children + indentation.
const treeChars = {
  vertical: '│',
  horizontal: '─',
  corner: '└',
  tee: '├',
  space: ' ',
};

function renderRange(range: RangeTableOfContentsNode | null, skipCanvases = false, indent = 0) {
  if (!range) {
    return '';
  }
  const spaces = treeChars.space.repeat(indent);
  let str = `${getValue(range.label)}\n`;
  const itemsCount = range.items ? range.items.length : 0;
  range.items?.forEach((item, index) => {
    const isLastItem = index === itemsCount - 1;
    if (item.isCanvasLeaf && skipCanvases) return;
    if (typeof item === 'string') {
      str += `${spaces}${isLastItem ? treeChars.corner : treeChars.tee}${treeChars.horizontal}${treeChars.horizontal} ${item}\n`;
    } else {
      str += `${spaces}${isLastItem ? treeChars.corner : treeChars.tee}${treeChars.horizontal}${treeChars.horizontal} ${renderRange(item, skipCanvases, indent + 2)}`;
    }
  });
  return str;
}

describe('range helper', () => {
  describe('rangeToTableOfContentsTree', () => {
    test('it can make a table of contents tree', () => {
      const vault = new Vault();
      const manifest = vault.loadSync<ManifestNormalized>(tableOfContentManifests.id, tableOfContentManifests);

      invariant(manifest);

      const tree = rangeToTableOfContentsTree(vault, manifest.structures[0] as any);

      expect(renderRange(tree)).toMatchInlineSnapshot(`
        "Table of Contents
        ├── Tabiba Tabiban [ጠቢበ ጠቢባን]
          ├── f. 1r
          └── f. 1v
        └── Arede'et [አርድዕት]
          ├── Monday
            ├── f. 2r
            └── f. 2v
          └── Tuesday
            ├── f. 3r
            └── f. 3v
        "
      `);
      //todo
      expect(tree).toMatchInlineSnapshot(`
        {
          "firstCanvas": {
            "selector": undefined,
            "source": {
              "id": "https://iiif.io/api/cookbook/recipe/0024-book-4-toc/canvas/p1",
              "type": "Canvas",
            },
            "type": "SpecificResource",
          },
          "id": "https://iiif.io/api/cookbook/recipe/0024-book-4-toc/range/r0",
          "isCanvasLeaf": false,
          "isRangeLeaf": false,
          "items": [
            {
              "firstCanvas": {
                "selector": undefined,
                "source": {
                  "id": "https://iiif.io/api/cookbook/recipe/0024-book-4-toc/canvas/p1",
                  "type": "Canvas",
                },
                "type": "SpecificResource",
              },
              "id": "https://iiif.io/api/cookbook/recipe/0024-book-4-toc/range/r1",
              "isCanvasLeaf": false,
              "isRangeLeaf": true,
              "items": [
                {
                  "id": "https://iiif.io/api/cookbook/recipe/0024-book-4-toc/canvas/p1",
                  "isCanvasLeaf": true,
                  "isRangeLeaf": false,
                  "label": {
                    "en": [
                      "f. 1r",
                    ],
                  },
                  "resource": {
                    "selector": undefined,
                    "source": {
                      "id": "https://iiif.io/api/cookbook/recipe/0024-book-4-toc/canvas/p1",
                      "type": "Canvas",
                    },
                    "type": "SpecificResource",
                  },
                  "type": "Canvas",
                  "untitled": false,
                },
                {
                  "id": "https://iiif.io/api/cookbook/recipe/0024-book-4-toc/canvas/p2",
                  "isCanvasLeaf": true,
                  "isRangeLeaf": false,
                  "label": {
                    "en": [
                      "f. 1v",
                    ],
                  },
                  "resource": {
                    "selector": undefined,
                    "source": {
                      "id": "https://iiif.io/api/cookbook/recipe/0024-book-4-toc/canvas/p2",
                      "type": "Canvas",
                    },
                    "type": "SpecificResource",
                  },
                  "type": "Canvas",
                  "untitled": false,
                },
              ],
              "label": {
                "gez": [
                  "Tabiba Tabiban [ጠቢበ ጠቢባን]",
                ],
              },
              "type": "Range",
              "untitled": false,
            },
            {
              "firstCanvas": {
                "selector": undefined,
                "source": {
                  "id": "https://iiif.io/api/cookbook/recipe/0024-book-4-toc/canvas/p3",
                  "type": "Canvas",
                },
                "type": "SpecificResource",
              },
              "id": "https://iiif.io/api/cookbook/recipe/0024-book-4-toc/range/r2",
              "isCanvasLeaf": false,
              "isRangeLeaf": false,
              "items": [
                {
                  "firstCanvas": {
                    "selector": undefined,
                    "source": {
                      "id": "https://iiif.io/api/cookbook/recipe/0024-book-4-toc/canvas/p3",
                      "type": "Canvas",
                    },
                    "type": "SpecificResource",
                  },
                  "id": "https://iiif.io/api/cookbook/recipe/0024-book-4-toc/range/r2/1",
                  "isCanvasLeaf": false,
                  "isRangeLeaf": true,
                  "items": [
                    {
                      "id": "https://iiif.io/api/cookbook/recipe/0024-book-4-toc/canvas/p3",
                      "isCanvasLeaf": true,
                      "isRangeLeaf": false,
                      "label": {
                        "en": [
                          "f. 2r",
                        ],
                      },
                      "resource": {
                        "selector": undefined,
                        "source": {
                          "id": "https://iiif.io/api/cookbook/recipe/0024-book-4-toc/canvas/p3",
                          "type": "Canvas",
                        },
                        "type": "SpecificResource",
                      },
                      "type": "Canvas",
                      "untitled": false,
                    },
                    {
                      "id": "https://iiif.io/api/cookbook/recipe/0024-book-4-toc/canvas/p4",
                      "isCanvasLeaf": true,
                      "isRangeLeaf": false,
                      "label": {
                        "en": [
                          "f. 2v",
                        ],
                      },
                      "resource": {
                        "selector": undefined,
                        "source": {
                          "id": "https://iiif.io/api/cookbook/recipe/0024-book-4-toc/canvas/p4",
                          "type": "Canvas",
                        },
                        "type": "SpecificResource",
                      },
                      "type": "Canvas",
                      "untitled": false,
                    },
                  ],
                  "label": {
                    "en": [
                      "Monday",
                    ],
                  },
                  "type": "Range",
                  "untitled": false,
                },
                {
                  "firstCanvas": {
                    "selector": undefined,
                    "source": {
                      "id": "https://iiif.io/api/cookbook/recipe/0024-book-4-toc/canvas/p5",
                      "type": "Canvas",
                    },
                    "type": "SpecificResource",
                  },
                  "id": "https://iiif.io/api/cookbook/recipe/0024-book-4-toc/range/r2/2",
                  "isCanvasLeaf": false,
                  "isRangeLeaf": true,
                  "items": [
                    {
                      "id": "https://iiif.io/api/cookbook/recipe/0024-book-4-toc/canvas/p5",
                      "isCanvasLeaf": true,
                      "isRangeLeaf": false,
                      "label": {
                        "en": [
                          "f. 3r",
                        ],
                      },
                      "resource": {
                        "selector": undefined,
                        "source": {
                          "id": "https://iiif.io/api/cookbook/recipe/0024-book-4-toc/canvas/p5",
                          "type": "Canvas",
                        },
                        "type": "SpecificResource",
                      },
                      "type": "Canvas",
                      "untitled": false,
                    },
                    {
                      "id": "https://iiif.io/api/cookbook/recipe/0024-book-4-toc/canvas/p6",
                      "isCanvasLeaf": true,
                      "isRangeLeaf": false,
                      "label": {
                        "en": [
                          "f. 3v",
                        ],
                      },
                      "resource": {
                        "selector": undefined,
                        "source": {
                          "id": "https://iiif.io/api/cookbook/recipe/0024-book-4-toc/canvas/p6",
                          "type": "Canvas",
                        },
                        "type": "SpecificResource",
                      },
                      "type": "Canvas",
                      "untitled": false,
                    },
                  ],
                  "label": {
                    "en": [
                      "Tuesday",
                    ],
                  },
                  "type": "Range",
                  "untitled": false,
                },
              ],
              "label": {
                "gez": [
                  "Arede'et [አርድዕት]",
                ],
              },
              "type": "Range",
              "untitled": false,
            },
          ],
          "label": {
            "en": [
              "Table of Contents",
            ],
          },
          "type": "Range",
          "untitled": false,
        }
      `);
    });

    test('it can make a table of contents tree', () => {
      const vault = new Vault();
      const manifest = vault.loadSync<ManifestNormalized>(tableOfContentManifests.id, tableOfContentsAvManifest);

      invariant(manifest);

      const tree = rangeToTableOfContentsTree(vault, manifest.structures[0] as any);

      expect(renderRange(tree, true)).toMatchInlineSnapshot(`
        "Gaetano Donizetti, L'Elisir D'Amore
        ├── Atto Primo
          ├── Preludio e Coro d'introduzione – Bel conforto al mietitore
          └── Remainder of Atto Primo
        └── Atto Secondo
        "
      `);

      expect(tree).toMatchInlineSnapshot(`
        {
          "firstCanvas": {
            "selector": {
              "type": "FragmentSelector",
              "value": "t=0,302.05",
            },
            "source": {
              "id": "https://iiif.io/api/cookbook/recipe/0026-toc-opera/canvas/1",
              "type": "Canvas",
            },
            "type": "SpecificResource",
          },
          "id": "https://iiif.io/api/cookbook/recipe/0026-toc-opera/range/1",
          "isCanvasLeaf": false,
          "isRangeLeaf": false,
          "items": [
            {
              "firstCanvas": {
                "selector": {
                  "type": "FragmentSelector",
                  "value": "t=0,302.05",
                },
                "source": {
                  "id": "https://iiif.io/api/cookbook/recipe/0026-toc-opera/canvas/1",
                  "type": "Canvas",
                },
                "type": "SpecificResource",
              },
              "id": "https://iiif.io/api/cookbook/recipe/0026-toc-opera/range/2",
              "isCanvasLeaf": false,
              "isRangeLeaf": false,
              "items": [
                {
                  "firstCanvas": {
                    "selector": {
                      "type": "FragmentSelector",
                      "value": "t=0,302.05",
                    },
                    "source": {
                      "id": "https://iiif.io/api/cookbook/recipe/0026-toc-opera/canvas/1",
                      "type": "Canvas",
                    },
                    "type": "SpecificResource",
                  },
                  "id": "https://iiif.io/api/cookbook/recipe/0026-toc-opera/range/3",
                  "isCanvasLeaf": false,
                  "isRangeLeaf": true,
                  "items": [
                    {
                      "id": "https://iiif.io/api/cookbook/recipe/0026-toc-opera/canvas/1#t=0,302.05",
                      "isCanvasLeaf": true,
                      "isRangeLeaf": false,
                      "label": {
                        "none": [
                          "Untitled",
                        ],
                      },
                      "resource": {
                        "selector": {
                          "type": "FragmentSelector",
                          "value": "t=0,302.05",
                        },
                        "source": {
                          "id": "https://iiif.io/api/cookbook/recipe/0026-toc-opera/canvas/1",
                          "type": "Canvas",
                        },
                        "type": "SpecificResource",
                      },
                      "type": "Canvas",
                      "untitled": true,
                    },
                  ],
                  "label": {
                    "it": [
                      "Preludio e Coro d'introduzione – Bel conforto al mietitore",
                    ],
                  },
                  "type": "Range",
                  "untitled": false,
                },
                {
                  "firstCanvas": {
                    "selector": {
                      "type": "FragmentSelector",
                      "value": "t=302.05,3971.24",
                    },
                    "source": {
                      "id": "https://iiif.io/api/cookbook/recipe/0026-toc-opera/canvas/1",
                      "type": "Canvas",
                    },
                    "type": "SpecificResource",
                  },
                  "id": "https://iiif.io/api/cookbook/recipe/0026-toc-opera/range/4",
                  "isCanvasLeaf": false,
                  "isRangeLeaf": true,
                  "items": [
                    {
                      "id": "https://iiif.io/api/cookbook/recipe/0026-toc-opera/canvas/1#t=302.05,3971.24",
                      "isCanvasLeaf": true,
                      "isRangeLeaf": false,
                      "label": {
                        "none": [
                          "Untitled",
                        ],
                      },
                      "resource": {
                        "selector": {
                          "type": "FragmentSelector",
                          "value": "t=302.05,3971.24",
                        },
                        "source": {
                          "id": "https://iiif.io/api/cookbook/recipe/0026-toc-opera/canvas/1",
                          "type": "Canvas",
                        },
                        "type": "SpecificResource",
                      },
                      "type": "Canvas",
                      "untitled": true,
                    },
                  ],
                  "label": {
                    "en": [
                      "Remainder of Atto Primo",
                    ],
                  },
                  "type": "Range",
                  "untitled": false,
                },
              ],
              "label": {
                "it": [
                  "Atto Primo",
                ],
              },
              "type": "Range",
              "untitled": false,
            },
            {
              "firstCanvas": {
                "selector": {
                  "type": "FragmentSelector",
                  "value": "t=3971.24",
                },
                "source": {
                  "id": "https://iiif.io/api/cookbook/recipe/0026-toc-opera/canvas/1",
                  "type": "Canvas",
                },
                "type": "SpecificResource",
              },
              "id": "https://iiif.io/api/cookbook/recipe/0026-toc-opera/range/5",
              "isCanvasLeaf": false,
              "isRangeLeaf": true,
              "items": [
                {
                  "id": "https://iiif.io/api/cookbook/recipe/0026-toc-opera/canvas/1#t=3971.24",
                  "isCanvasLeaf": true,
                  "isRangeLeaf": false,
                  "label": {
                    "none": [
                      "Untitled",
                    ],
                  },
                  "resource": {
                    "selector": {
                      "type": "FragmentSelector",
                      "value": "t=3971.24",
                    },
                    "source": {
                      "id": "https://iiif.io/api/cookbook/recipe/0026-toc-opera/canvas/1",
                      "type": "Canvas",
                    },
                    "type": "SpecificResource",
                  },
                  "type": "Canvas",
                  "untitled": true,
                },
              ],
              "label": {
                "it": [
                  "Atto Secondo",
                ],
              },
              "type": "Range",
              "untitled": false,
            },
          ],
          "label": {
            "it": [
              "Gaetano Donizetti, L'Elisir D'Amore",
            ],
          },
          "type": "Range",
          "untitled": false,
        }
      `);
    });

    test('it can make a table of contents tree', () => {
      const vault = new Vault();
      const manifest = vault.loadSync<ManifestNormalized>(wellcomeRange['@id'], wellcomeRange);

      invariant(manifest);

      const tree = rangesToTableOfContentsTree(vault, manifest.structures);

      expect(renderRange(tree, true)).toMatchInlineSnapshot(`
        "Table of Contents
        ├── Cover
        ├── 'Precatio terrae matris,' abbreviated and corrupted
        ├── 'Precatio omnium herbarum,' abbreviated and corrupted
        ├── Antonius Musa, 'De herba vettonica'
        ├── Pseudo-Apuleius, 'Herbarius'
        ├── Anonymous, 'De taxone'
        ├── Sextus Placitus, 'De medicina ex animalibus,' (continues at f.76r)
        ├── 'Curae herbarum,' interpolated within the Sextus Placitus herbal
        ├── Receipts, interpolated within the Sextus Placitus herbal
        ├── Illustration
        ├── Sextus Placitus, 'De medicina ex animalibus,' (continued from f.46v)
        ├── Relation of a cure of a friend of the scribe
        ├── Tables of Roman and Arabic numerals from 1 to 1,000,000
        └── Cover
        "
      `);
    });

    test('it can parse BL manifest', () => {
      const vault = new Vault();
      const manifest = vault.loadSync<ManifestNormalized>(blManifest.id, blManifest);

      invariant(manifest);

      const tree = rangesToTableOfContentsTree(vault, manifest.structures);

      expect(renderRange(tree, true)).toMatchInlineSnapshot(`
        "Wiltshire and Dorset, dub of disks / Fanny Rumble, A. Collins, Perrier (01:07:55)
        ├── The turmut hoeing (02:29)
        ├── She stole my heart away (02:08)
        ├── Dumble dum dollicky (Richard of Taunton Dean) (03:01)
        ├── Mrs Fanny Rumble talks about herself (01:44)
        ├── What shall I wear to the wedding, John? (03:25)
        ├── Country courtship (05:50)
        ├── Herbert Prince (05:00)
          ├── Introductory talk: 'The young sailor cut down in his prime' (01:25)
          └── The young sailor cut down in his prime (02:49)
        └── Fanny Rumble / Albert Collins / Fred Perrier (25:17)
          ├── O what shall I wear to the wedding, John? (03:57)
          ├── O what shall I wear to the wedding, John? (02:47)
          ├── The vly on the turmut (03:10)
          ├── The vly on the turmut (01:37)
          ├── Twas on a Monday morning (02:04)
          ├── Twas on a Monday morning (02:22)
          ├── Dumble dum dollicky (04:23)
          └── Talk about herself (01:53)
        "
      `);
    });
  });
});
