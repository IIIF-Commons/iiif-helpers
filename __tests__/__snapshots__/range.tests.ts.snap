// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`range helper > it can parse bodleian ranges 2`] = `
[
  {
    "accompanyingCanvas": null,
    "annotations": [],
    "behavior": [
      "top",
    ],
    "homepage": [],
    "id": "https://iiif.bodleian.ox.ac.uk/iiif/range/390fd0e8-9eae-475d-9564-ed916ab9035c/LOG_0000",
    "items": [
      {
        "id": "https://iiif.bodleian.ox.ac.uk/iiif/range/390fd0e8-9eae-475d-9564-ed916ab9035c/LOG_0001",
        "type": "Range",
      },
      {
        "id": "https://iiif.bodleian.ox.ac.uk/iiif/range/390fd0e8-9eae-475d-9564-ed916ab9035c/LOG_0011",
        "type": "Range",
      },
      {
        "id": "https://iiif.bodleian.ox.ac.uk/iiif/range/390fd0e8-9eae-475d-9564-ed916ab9035c/LOG_0186",
        "type": "Range",
      },
      {
        "id": "https://iiif.bodleian.ox.ac.uk/iiif/range/390fd0e8-9eae-475d-9564-ed916ab9035c/LOG_0321",
        "type": "Range",
      },
      {
        "id": "https://iiif.bodleian.ox.ac.uk/iiif/range/390fd0e8-9eae-475d-9564-ed916ab9035c/LOG_0454",
        "type": "Range",
      },
      {
        "selector": undefined,
        "source": {
          "id": "https://iiif.bodleian.ox.ac.uk/iiif/canvas/57e041f5-8945-44b6-918d-b949e5c97ba1.json",
          "type": "Canvas",
        },
        "type": "SpecificResource",
      },
      {
        "selector": undefined,
        "source": {
          "id": "https://iiif.bodleian.ox.ac.uk/iiif/canvas/a13ee009-596c-4151-8f4c-568b0c2ce2cc.json",
          "type": "Canvas",
        },
        "type": "SpecificResource",
      },
      {
        "selector": undefined,
        "source": {
          "id": "https://iiif.bodleian.ox.ac.uk/iiif/canvas/4534b928-7fde-443a-8664-362b260ac72e.json",
          "type": "Canvas",
        },
        "type": "SpecificResource",
      },
      {
        "selector": undefined,
        "source": {
          "id": "https://iiif.bodleian.ox.ac.uk/iiif/canvas/9c33c6c5-1f11-4465-92a4-eb199dd490aa.json",
          "type": "Canvas",
        },
        "type": "SpecificResource",
      },
      {
        "selector": undefined,
        "source": {
          "id": "https://iiif.bodleian.ox.ac.uk/iiif/canvas/7ecb129e-0e2c-4bbd-9815-30c3c60cc86d.json",
          "type": "Canvas",
        },
        "type": "SpecificResource",
      },
      {
        "selector": undefined,
        "source": {
          "id": "https://iiif.bodleian.ox.ac.uk/iiif/canvas/81bd489f-5560-434b-a982-720f987a69c5.json",
          "type": "Canvas",
        },
        "type": "SpecificResource",
      },
      {
        "selector": undefined,
        "source": {
          "id": "https://iiif.bodleian.ox.ac.uk/iiif/canvas/2e9a76be-2fc1-45c0-84c9-2e47465e7424.json",
          "type": "Canvas",
        },
        "type": "SpecificResource",
      },
      {
        "selector": undefined,
        "source": {
          "id": "https://iiif.bodleian.ox.ac.uk/iiif/canvas/40fdea5a-1aad-4085-8e68-857c6dbbc414.json",
          "type": "Canvas",
        },
        "type": "SpecificResource",
      },
      {
        "selector": undefined,
        "source": {
          "id": "https://iiif.bodleian.ox.ac.uk/iiif/canvas/a945d0ff-d18a-49fd-93ac-f53118fbd606.json",
          "type": "Canvas",
        },
        "type": "SpecificResource",
      },
      {
        "selector": undefined,
        "source": {
          "id": "https://iiif.bodleian.ox.ac.uk/iiif/canvas/c1d4f356-977c-4e2e-8788-3ca65f3815a0.json",
          "type": "Canvas",
        },
        "type": "SpecificResource",
      },
      {
        "selector": undefined,
        "source": {
          "id": "https://iiif.bodleian.ox.ac.uk/iiif/canvas/2d886d2c-7b06-4da9-b527-c7a8d3b6751b.json",
          "type": "Canvas",
        },
        "type": "SpecificResource",
      },
      {
        "selector": undefined,
        "source": {
          "id": "https://iiif.bodleian.ox.ac.uk/iiif/canvas/d994b374-3fb3-4c2f-b8b5-a53d1f6363e1.json",
          "type": "Canvas",
        },
        "type": "SpecificResource",
      },
      {
        "selector": undefined,
        "source": {
          "id": "https://iiif.bodleian.ox.ac.uk/iiif/canvas/e2a05495-79d5-4300-ba58-63ead7a3768b.json",
          "type": "Canvas",
        },
        "type": "SpecificResource",
      },
      {
        "selector": undefined,
        "source": {
          "id": "https://iiif.bodleian.ox.ac.uk/iiif/canvas/16cc93d2-c242-4f30-9536-b5c05a4dffdd.json",
          "type": "Canvas",
        },
        "type": "SpecificResource",
      },
    ],
    "label": {
      "none": [
        "Arch. G c.7",
      ],
    },
    "metadata": [],
    "navDate": null,
    "partOf": [],
    "placeholderCanvas": null,
    "provider": [],
    "rendering": [],
    "requiredStatement": null,
    "rights": null,
    "seeAlso": [],
    "service": [],
    "start": {
      "id": "https://iiif.bodleian.ox.ac.uk/iiif/canvas/57e041f5-8945-44b6-918d-b949e5c97ba1.json",
      "type": "Canvas",
    },
    "summary": null,
    "supplementary": null,
    "thumbnail": [],
    "type": "Range",
    "viewingDirection": "left-to-right",
  },
]
`;
